### PR TITLE
feat: Add `NullableFromType` option

### PR DESCRIPTION
**Pull request by candiduslynx**
Instead of https://github.com/invopop/jsonschema/pull/106.

Allows to define property nullability based on the field type instead of the `jsonschema:"nullable"` tag.
Useful when the types included are referenced from 3rd party packages.

Consider the following example:
```go
type TestNullableField struct {
	A *string `json:"a"`
}
```

The following JSON is a perfectly valid value for Go:
```json
{
  "a": null
}
```

The schemas produced for the `TestNullableField` struct are:

<details><summary>Old behavior (`NullableFromType:false`)</summary>
<p>

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://github.com/invopop/jsonschema/test-nullable-field",
  "$ref": "#/$defs/TestNullableField",
  "$defs": {
    "TestNullableField": {
      "properties": {
        "a": {
          "type": "string"
        }
      },
      "additionalProperties": false,
      "type": "object",
      "required": [
        "a"
      ]
    }
  }
}
```

</p>
</details>

<details><summary>New behavior (`NullableFromType:true`)</summary>
<p>

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://github.com/invopop/jsonschema/test-nullable-field",
  "$ref": "#/$defs/TestNullableField",
  "$defs": {
    "TestNullableField": {
      "properties": {
        "a": {
          "oneOf": [
            {
              "type": "string"
            },
            {
              "type": "null"
            }
          ]
        }
      },
      "additionalProperties": false,
      "type": "object",
      "required": [
        "a"
      ]
    }
  }
}
```

</p>
</details>

In both cases `a` is required, but the old behavior will fail for `{"a":null}` whereas the new one will succeed.

### CODE REVIEW

**Comment by candiduslynx** [comment]
@samlown this is a less intrusive implementation for https://github.com/invopop/jsonschema/pull/106.
It will kick in the new logic only if the `NullableFromType` option is used.

**Comment by kop** [comment]
Highly interested in this PR as well. Is there anything I can help with to get this merged?

**Comment by candiduslynx** [comment]
> Highly interested in this PR as well. Is there anything I can help with to get this merged?

@kop I really hope @samlown has time to review/merge that.
In the meantime, we use a fork at github.com/cloudquery/jsonschema. It has `cqmain` default branch, that has the following PRs merged:
* https://github.com/invopop/jsonschema/pull/109
* https://github.com/invopop/jsonschema/pull/110
* https://github.com/invopop/jsonschema/pull/111
* https://github.com/invopop/jsonschema/pull/112

File: reflect.go [hunk 1]
```diff
@@ -968,6 +981,19 @@ func nullableFromJSONSchemaTags(tags []string) bool {
...
 }
 
+func isNullable(k reflect.Kind) bool {
+	switch k {
+	case reflect.Pointer,
+		reflect.UnsafePointer,
+		reflect.Map,
+		reflect.Slice,
+		reflect.Interface:
+		return true
```
Comment (samlown) [inline reflect.go:991]: I do think there should be specific tests for these types. I also wonder if `Interface` and `UnsafePointer` are really necessary, if so, there should be a test to show it ;-)

**Review by samlown — COMMENTED** [review]
Sorry for the delay there! This looks good to me, thanks! I do think there should be some extra tests for those non-pointer reflection types though.

File: reflect.go (hunk 1 above)
Comment (candiduslynx) [reply reflect.go:997]: @samlown I've added the tests in be77f3d54585a05a330261a21c6dce7bac4d08bb

I can see the appeal to remove both Interface & UnsafePointer as well, so I can update the code accordingly, if you insist.

I also updated the code to actually handle a more varied cases like `map[string]*int` or `*[]*int` & included that in tests.

**Comment by candiduslynx** [comment]
@samlown thanks for the review!

I have an additional question here: currently the library uses `oneOf` to make schema nullable.
In my opinion, a better way would be `anyOf`:
If the schema is already nullable (e.g., it already is `{"oneOf":[{...}, {"type":"null"}]}`) then adding another `oneOf` layer will effectively make `null` value to fail validation, as the `oneOf` requires one & only one variant to be valid:
```json
{
  "oneOf": [
    {
      "oneOf": [
        {},
        {
          "type": "null"
        }
      ]
    },
    {
      "type": "null"
    }
  ]
}
```
If we use `anyOf` this is lifted to `at least one of the variants`.

I can raise this as a separate issue & address it in a follow-up PR.
WDYT?

**Comment by Rosswell** [comment]
@samlown would you be able to take another look when you have a chance?

**Comment by candiduslynx** [comment]
Hi @samlown!

Could you please advice if you need any assistance in maintenance of this library?

I see that the last commit here was 4 months ago, & I wonder whether this library is actually actively maintained or not.

File: reflect.go [hunk 2]
```diff
@@ -1023,7 +1055,12 @@ func (r *Reflector) reflectFieldName(f reflect.StructField) (string, bool, bool,
 	}
 	requiredFromJSONSchemaTags(schemaTags, &required)
 
-	nullable := nullableFromJSONSchemaTags(schemaTags)
+	var nullable bool
+	if r.NullableFromType {
+		nullable = isNullable(f.Type.Kind())
+	} else {
+		nullable = nullableFromJSONSchemaTags(schemaTags)
+	}
```
Comment (samlown) [inline reflect.go:1063]: It feels like this logic should be inside its own method.

File: reflect.go [hunk 3]
```diff
@@ -511,23 +533,20 @@ func (r *Reflector) reflectStructFields(st *Schema, definitions Definitions, t r
 			property = r.refOrReflectTypeToSchema(definitions, f.Type)
 		}
 
-		property.structKeywordsFromTags(f, st, name)
-		if property.Description == "" {
-			property.Description = r.lookupComment(t, f.Name)
+		unwrapped := property
```
Comment (samlown) [inline reflect.go:536]: What is the objective here? I can't see where `unwrapped` is consumed, making this section much harder to read as I'd guess everything is happening on the pointer.

**Review by samlown — CHANGES_REQUESTED** [review]
Sorry for the delay, I really don't have much time at the moment. There are quite a few style things here that I'd be worried about, especially regarding public methods which are not tested (just make them private!)

File: reflect.go (hunk 3 above)
Comment (candiduslynx) [reply reflect.go:536]: https://github.com/invopop/jsonschema/pull/110/files#diff-672d1dd2b9193c8818108f625c29ff25feaea4c1c50cc36cfda18108cba3e1b2R540
We need this to work with pointer fields as well

**Comment by candiduslynx** [comment]
@samlown I'm closing this PR for 2 reasons:
1. I no longer have access to the fork the PR was opened from
2. It seems to me that you just don't have the capacity to get this reviewed and merged

If you'll be available for review & merging please tag me here & I'll reopen from my own fork.

**Pull request closed without merging**